### PR TITLE
ngcc: add build marker

### PR DESCRIPTION
**Pull request by petebacondarwin**
This sits on top of #25406

### CODE REVIEW

**Review by gkalpak — COMMENTED** [review]
AFAICT, more logic is needed. E.g. when we detect that a package has been compiled by a different ngcc version. 
I.e. we need a way to "uncompile" before compiling with the new version. ("Uncompiling" could be as simple as rimrafing `node_modules_ngtsc/` or as complex as walking `node_modules/` and renaming all backup files to their original name.)

I think we also need to record the compilations output path (e.g. `node_modules/` or `node_modules_ngtsc/`) in the marker metadata.

File: packages/compiler-cli/src/ngcc/src/transform/utils.ts [hunk 1]
```diff
@@ -6,12 +6,14 @@
...
 
-import {readFileSync} from 'fs';
-import {dirname, relative, resolve} from 'path';
+import {existsSync, readFileSync, writeFileSync} from 'fs';
+import {dirname, resolve} from 'path';
 import {find} from 'shelljs';
 
 import {isDefined} from '../utils';
 
+export const NGCC_VERSION = '0.0.0-PLACEHOLDER';
```
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngcc/src/transform/utils.ts]: I would rather not have a dependency on compiler-cli if we can avoid it. That version is created in exactly the same way as this anyway.

**Comment by petebacondarwin** [comment]
We should avoid trying to "uncompile" at this stage. The backups are there if we need them for adding such a strategy later. In the meantime I will add a message telling the user to blow away their node_modules if the ngcc version changes.

File: packages/compiler-cli/src/ngcc/src/transform/utils.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (gkalpak) [reply packages/compiler-cli/src/ngcc/src/transform/utils.ts]: Hm...I thought `ngcc` will be part of the `compiler-cli` package anyway.
In any case, not a big deal :grin:

[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngcc/src/transform/utils.ts]: To give more strength to this approach I can also say that I conferred with Alex Eagle about the best approach here as well :-)

**Comment by petebacondarwin** [comment]
We just need to work out what to do about the failing integration test: bazel tests do not allow the subject to write over the node_modules of the test folder.

**Review by alxhub — APPROVED** [review]

**Comment by petebacondarwin** [comment]
I have implemented an (in-memory) mock filesystem for the integration test that was failing.
This allows the test to be able to write to the `node_modules` folder safely. It could also allow
us to do more subtle testing of file changes at this level, going forward.

**Comment by petebacondarwin** [comment]
Awaiting G3 resubmit and pullapprove for `integration` test.
^^^ @alxhub or @mhevery

**Pull request closed without merging**